Update completion for tmutil

samdoran opened the pull request:
## Description

- Reorganize completions and options so they are easier to read.
- Add destination UUIDs to completetions as well as descriptions
- Add a few missing sub commands

The biggest change is adding the destination UUIDs to the completion. This is very helpful with `tmutil startbackup -d [destination ID]` or `tmutil setquota [destination ID]`.

I used the neat feature of `fish` to dynamically populate the completion description as well as the value.

One thing I could use help with is how to make the completion values update dynamically when destinations are added/removed. Currently I have to source the completions again to update the destination IDs. I know there is a way to do this but I couldn't quite figure it out. Thanks!

Fixes issue #

## TODOs:
<!-- Just check off what what we know been done so far. We can help you with this stuff. -->
- [ ] Changes to fish usage are reflected in user documentation/manpages.
- [ ] Tests have been added for regressions fixed
- [ ] User-visible changes noted in CHANGELOG.rst <!-- Don't document changes for completions inside CHANGELOG.rst, there are lot of such edits -->

samdoran committed:
Update completion for tmutil

- Reorganize completions and options so they are easier to read.
- Add destination UUIDs to completetions as well as descriptions
- Add a few missing sub commands

faho commented on share/completions/tmutil.fish:
```diff
@@ -1,61 +1,136 @@
-complete -f -c tmutil -n '__fish_seen_subcommand_from destinationinfo isexcluded compare' -s X -d 'Print as XML'
+function __destination_ids
+    for line in $(tmutil destinationinfo)
+        if test -n "$(string match '*===*' $line)"
```
`string match -q`, no `test`.

Also use `--` before the variable, or it'll break on lines starting with `-`.

So:

```fish
if string match -q '*===*' -- $line
```

Same goes for the other `string` uses.

samdoran replied:
Oh that's much nicer. Thanks!

faho commented on share/completions/tmutil.fish:
```diff
@@ -1,61 +1,136 @@
+        if test -n "$(string match '*===*' $line)"
+            # New section so clear out variables
+            set -f name ''
+            set -f ID ''
```
ID doesn't need to be a separate variable here because it's never used without being printed.

samdoran replied:
This was extra diligence to make sure a value for ID from a previous section didn't get reused, but I guess that's unnecessary.

faho commented on share/completions/tmutil.fish:
```diff
@@ -1,61 +1,136 @@
+complete -c tmutil -f -n __fish_use_subcommand -a machinedirectory -d 'Print the path to the current machine directory'
+
+complete -c tmutil -f -n __fish_use_subcommand -a removedestination -d 'Removes a backup destination'
+complete -c tmutil -f -n '__fish_seen_subcommand_from removedestination' -d "$(__destination_ids)"
```
You want this to be an *argument*, not a description, presumably? I.e. `-a` instead of `-d`?

samdoran replied:
Yes. I corrected one of those but I missed one. Thank you for catching it.

faho commented on share/completions/tmutil.fish:
```diff
@@ -1,61 +1,136 @@
+complete -c tmutil -f -n '__fish_seen_subcommand_from startbackup' -s a -l auto -d 'Automatic mode'
+complete -c tmutil -f -n '__fish_seen_subcommand_from startbackup' -s b -l block -d 'Block until finished'
+complete -c tmutil -f -n '__fish_seen_subcommand_from startbackup' -s r -l rotation -d 'Automatic rotation'
+complete -c tmutil -f -n '__fish_seen_subcommand_from startbackup' -s d -l destination -r -a "$(__destination_ids)" -d 'Backup destination'
```
This:

```fish
"$(...)"
```

is a command substitution that will be expanded when this is sourced.

What you want is to use `-a`'s ability to do that when the argument is needed:

```fish
-a "(__destination_ids)"
```

(or `'$(__destination_ids)'` - the important part is that the command substitution is quoted so it doesn't run when the `complete` command is run, but when the `-a` is evaluated)

samdoran replied:
That's brilliant! Essentially avoid expanding the command substitution within the quotes and it gets run later.

This works as expected in my testing.

samdoran committed:
Do not use test for evaluating string match

Also add `--` to ensure parameters don't get mixed up with the line itself.

samdoran committed:
Argument not description

samdoran committed:
No need to reset ID

The values are only echoed if ID matches, so it will always be correct.

samdoran committed:
Evaluate completion arguments when called

This makes the destinations update dynamically when they are added/removed.
Unquote the echo statement so that it is correctly paresd by the -a options.

samdoran commented:
I've got a few more subtle behaviors to iron out mainly around when to offer file completion for certain subcommands. I think the `-F` use is incorrect.

samdoran committed:
Refine when file completions are offered and add a missing short arg

samdoran commented:
This should be good to merge now. Thanks for the very helpful review.

samdoran commented:
Thank you for merging and for the great review.
